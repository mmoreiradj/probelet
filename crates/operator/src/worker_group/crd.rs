use std::{
    collections::{BTreeMap, HashMap},
    sync::Arc,
    time::Duration,
};

use kube::{CustomResource, ResourceExt, runtime::controller::Action};
use schemars::JsonSchema;
use serde::{Deserialize, Serialize};

use super::Result;
use crate::{Context, metrics::MetricLabel, worker_group::reconcile::ReconcileWorkerGroupTask};

/// The `WorkerGroup` is a resource that manages a group of `Worker` instances (Pods).
/// `Workers` are where the probes are going to be executed.
#[derive(CustomResource, Deserialize, Serialize, Clone, Debug, JsonSchema)]
#[kube(
    kind = "WorkerGroup",
    group = "probelet.dev",
    version = "v0",
    namespaced
)]
#[kube(status = "WorkerGroupStatus", shortname = "workergroup")]
pub struct WorkerGroupSpec {
    /// The number of replicas to create
    /// The max is `1` for the first version of the CRD
    pub replicas: i32,
    /// The image to use for the `WorkerGroup`
    pub image: String,
}

#[derive(Deserialize, Serialize, Clone, Debug, JsonSchema, PartialEq, Eq, Hash)]
pub struct WorkerInstanceName(pub String);

impl TryFrom<String> for WorkerInstanceName {
    type Error = String;

    /// Try to convert a string to a `WorkerInstanceName`
    /// `WorkerInstanceName` has to be a valid kubernetes object name
    fn try_from(value: String) -> std::result::Result<Self, Self::Error> {
        if value.is_empty() {
            return Err("Worker instance name cannot be empty".to_string());
        }

        if value.len() > 63 {
            return Err("Worker instance name cannot be longer than 63 characters".to_string());
        }

        if value.contains('.') || value.contains('/') || value.contains('%') {
            return Err("Worker instance name cannot contain . or / or %".to_string());
        }

        Ok(WorkerInstanceName(value))
    }
}

impl WorkerInstanceName {
    pub fn as_str(&self) -> &str {
        &self.0
    }

    pub fn as_string(&self) -> String {
        self.0.clone()
    }
}

#[derive(Deserialize, Serialize, Clone, Debug, JsonSchema, PartialEq, Eq, Hash)]
pub enum WorkerGroupInstanceStatus {
    /// The instance is ready
    Ready,
    /// The instance is not ready
    NotReady,
}

#[derive(Deserialize, Serialize, Clone, Debug, JsonSchema, PartialEq, Eq, Hash)]
pub struct WorkerGroupReportedInstanceState {
    /// The state of the instance, can be `Ready` or `NotReady`
    pub status: WorkerGroupInstanceStatus,
    /// The last updated time of the instance
    pub last_updated: String,
}

/// The status object of `WorkerGroup`
#[derive(Deserialize, Serialize, Clone, Debug, JsonSchema)]
pub struct WorkerGroupStatus {
    /// Runnning instances names
    pub instance_names: Vec<WorkerInstanceName>,
    /// Number of instances
    pub instances: i32,
    /// State of instances
    pub instances_reported_state: HashMap<WorkerInstanceName, WorkerGroupReportedInstanceState>,
    /// Ready instances
    pub ready_instances: i32,
}

impl MetricLabel for WorkerGroup {
    fn metric_label(&self) -> String {
        format!("worker_group__{}", self.name_any())
    }
}

impl WorkerGroup {
    pub(crate) async fn reconcile(&self, context: Arc<Context>) -> Result<Action> {
        match ReconcileWorkerGroupTask::from_worker_group(self.clone(), context)? {
            Some(task) => task.run().await,
            None => Ok(Action::requeue(Duration::from_secs(5 * 60))),
        }
    }

    pub(crate) async fn cleanup(&self, _context: Arc<Context>) -> Result<Action> {
        Ok(Action::requeue(Duration::from_secs(5 * 60)))
    }

    pub fn default_annotations(&self) -> BTreeMap<String, String> {
        let mut annotations = BTreeMap::new();
        annotations.insert(
            "probelet.dev/operatorVersion".to_string(),
            env!("CARGO_PKG_VERSION").to_string(),
        );
        annotations
    }

    pub fn default_labels(&self) -> BTreeMap<String, String> {
        let mut labels = BTreeMap::new();
        labels.insert("probelet.dev/workerGroupName".to_string(), self.name_any());
        labels
    }
}
