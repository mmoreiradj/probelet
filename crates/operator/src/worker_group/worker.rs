use std::{sync::Arc, time::Duration};

use k8s_openapi::api::core::v1::{Container, Pod, PodSpec};
use kube::{
    Api, Resource, ResourceExt,
    api::{ObjectMeta, PostParams},
    runtime::{
        controller::Action,
        events::{Event, EventType},
    },
};
use snafu::ResultExt;

use crate::{
    Context,
    worker_group::{WorkerGroup, error::KubeSnafu, reconcile::EventReason},
};

use super::Result;

const WORKER_GROUP_DEFAULT_DELETION_GRACE_PERIOD_SECONDS: i64 = 30; // 30 seconds

#[derive(Debug, Clone)]
pub struct Worker {
    pub name: String,
    pub image: String,
    pub worker_group: Arc<WorkerGroup>,
}

impl Worker {
    pub fn new(name: String, image: String, worker_group: Arc<WorkerGroup>) -> Self {
        Self {
            name,
            image,
            worker_group,
        }
    }

    pub async fn create(&self, context: Arc<Context>) -> Result<Action> {
        let pod = self.pod();
        let client = context.client.clone();
        let api = Api::<Pod>::namespaced(client, &self.worker_group.namespace().unwrap());
        api.create(&PostParams::default(), &pod)
            .await
            .context(KubeSnafu {
                message: format!("Failed to create worker pod {}", self.name),
            })?;

        let event = Event {
            type_: EventType::Normal,
            reason: EventReason::WorkerCreated.to_string(),
            note: Some("Worker Created".to_string()),
            secondary: Some(self.worker_group.object_ref(&())),
            action: EventReason::WorkerCreated.to_string(),
        };

        let recorder = context.recorder.clone();
        recorder
            .publish(&event, &pod.object_ref(&()))
            .await
            .context(KubeSnafu {
                message: format!("Failed to publish event for worker {}", self.name),
            })?;

        Ok(Action::requeue(Duration::from_secs(5 * 60)))
    }

    pub fn pod(&self) -> Pod {
        let spec = PodSpec {
            containers: vec![Container {
                name: "worker".to_string(),
                image: Some(self.image.clone()),
                command: Some(vec!["/bin/sh".to_string(), "-c".to_string()]),
                args: Some(vec!["sleep infinity".to_string()]),
                ..Default::default()
            }],
            restart_policy: Some("Always".to_string()),
            ..Default::default()
        };

        let mut annotations = self.worker_group.default_annotations();
        annotations.insert(
            "probelet.dev/podSpec".to_string(),
            serde_json::to_string(&spec).unwrap(),
        );
        let mut labels = self.worker_group.default_labels();
        labels.insert("probelet.dev/workerName".to_string(), self.name.clone());

        Pod {
            metadata: ObjectMeta {
                name: Some(self.name.clone()),
                owner_references: Some(vec![self.worker_group.owner_ref(&()).unwrap()]),
                annotations: Some(annotations),
                deletion_grace_period_seconds: Some(
                    WORKER_GROUP_DEFAULT_DELETION_GRACE_PERIOD_SECONDS,
                ),
                labels: Some(labels),
                ..Default::default()
            },
            spec: Some(spec),
            ..Default::default()
        }
    }
}

#[cfg(test)]
mod tests {
    use insta::assert_snapshot;

    use crate::worker_group::crd::WorkerGroupSpec;

    use super::*;

    #[test_log::test(tokio::test)]
    async fn test_pod() {
        let worker = Worker::new(
            "test".to_string(),
            "test".to_string(),
            Arc::new(WorkerGroup {
                metadata: ObjectMeta {
                    name: Some("test".to_string()),
                    uid: Some("test".to_string()),
                    ..Default::default()
                },
                spec: WorkerGroupSpec {
                    replicas: 1,
                    image: "test".to_string(),
                },
                status: None,
            }),
        );

        let pod = worker.pod();
        let pod_json = serde_json::to_string_pretty(&pod).unwrap();

        assert_snapshot!(pod_json);
    }
}
