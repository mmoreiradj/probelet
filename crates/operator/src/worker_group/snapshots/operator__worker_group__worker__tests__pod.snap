---
source: crates/operator/src/worker_group/worker.rs
expression: pod_json
---
{
  "apiVersion": "v1",
  "kind": "Pod",
  "metadata": {
    "annotations": {
      "probelet.dev/operatorVersion": "0.1.0",
      "probelet.dev/podSpec": "{\"containers\":[{\"args\":[\"sleep infinity\"],\"command\":[\"/bin/sh\",\"-c\"],\"image\":\"test\",\"name\":\"worker\"}],\"restartPolicy\":\"Always\"}"
    },
    "deletionGracePeriodSeconds": 30,
    "labels": {
      "probelet.dev/workerGroupName": "test",
      "probelet.dev/workerName": "test"
    },
    "name": "test",
    "ownerReferences": [
      {
        "apiVersion": "probelet.dev/v0",
        "kind": "WorkerGroup",
        "name": "test",
        "uid": "test"
      }
    ]
  },
  "spec": {
    "containers": [
      {
        "args": [
          "sleep infinity"
        ],
        "command": [
          "/bin/sh",
          "-c"
        ],
        "image": "test",
        "name": "worker"
      }
    ],
    "restartPolicy": "Always"
  }
}
